src/utils/event/highlight.ts

```diff
@@ -0,0 +1,108 @@
+): boolean => {
+    return getEventHighlightInfo(event, client, timelineRenderingType).isHighlighted;
+};
+
+/**
+ * Determine whether an event should be highlighted and why
+ * For edited events, if a previous version of the event was highlighted
+ * the event should remain highlighted as the user may have been notified
+ * @returns {boolean} isHighlighted
+ * @returns {boolean} isBecausePreviousEvent - true when a pervious version of the event triggered a highlight
+ *                      but the current replaced event does not
+ * @returns {IAnnotatedPushRule} rule - the rule that triggered the highlight
+ */
+export const getEventHighlightInfo = (
+    event: MatrixEvent,
+    client: MatrixClient,
+    timelineRenderingType: TimelineRenderingType,
+): {
+    isHighlighted: boolean;
+    isBecausePreviousEvent?: boolean;
+    rule?: IAnnotatedPushRule;
+} => {
+    if (timelineRenderingType === TimelineRenderingType.Notification) return { isHighlighted: false };
+    if (timelineRenderingType === TimelineRenderingType.ThreadsList) return { isHighlighted: false };
+
+    // don't show self-highlights from another of our clients
+    if (event.getSender() === client.getSafeUserId()) {
+        return { isHighlighted: false };
+    }
+
+    const pushDetails = client.getPushDetailsForEvent(event.replacingEvent() || event);
+    // get the actions for the previous version of the event too if it is an edit
+    const previousPushDetails = event.replacingEvent() ? client.getPushDetailsForEvent(event) : null;
+
+    if (!pushDetails?.actions?.tweaks && !previousPushDetails?.actions?.tweaks) {
+        return { isHighlighted: false };
+    }
+
+    // current version of the event triggered a highlight
+    if (pushDetails?.actions?.tweaks?.highlight) {
+        return {
+            isHighlighted: true,
+            rule: pushDetails.rule,
+        };
+    }
+
+    // previous version of the event triggered a highlight
+    if (previousPushDetails?.actions?.tweaks.highlight) {
+        return {
+            isHighlighted: true,
+            rule: previousPushDetails.rule,
+            isBecausePreviousEvent: true,
+        };
+    }
+
+    return { isHighlighted: false };
+};
+
+export const getHighlightReasonMessage = (rule: IAnnotatedPushRule): string => {
+    if (rule.rule_id === RuleId.ContainsUserName) {
+        return _t("Your username was mentioned.");
+    }
+    if (rule.rule_id === RuleId.ContainsDisplayName) {
+        return _t("Your display name was mentioned.");
+    }
+    if (rule.rule_id === RuleId.AtRoomNotification || rule.rule_id === RuleId.IsRoomMention) {
+        return _t("This message mentions the room.");
+    }
+    if (rule.rule_id === RuleId.IsUserMention) {
+        return _t("You were explicitly mentioned.");
+    }
+    if (rule.kind === PushRuleKind.ContentSpecific) {
+        return _t(`Your keyword '%(keyword)s' was mentioned.`, { keyword: rule.pattern! });
+    }
+    return "";
```

Comment:
This should probably be a switch-case, at least [according to the code style](https://github.com/vector-im/element-web/blob/develop/code_style.md?plain=1#L127)